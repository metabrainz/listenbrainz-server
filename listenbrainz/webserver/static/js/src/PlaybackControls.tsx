import * as React from "react";
import {
  faEye,
  faEyeSlash,
  faFastBackward,
  faFastForward,
  faPauseCircle,
  faPlayCircle,
  faSortAmountDown,
  faSortAmountUp,
} from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconDefinition } from "@fortawesome/fontawesome-common-types"; // eslint-disable-line import/no-unresolved
import { IconProp } from "@fortawesome/fontawesome-svg-core";

type PlaybackControlsProps = {
  playPreviousTrack: () => void;
  playNextTrack: (invert?: boolean) => void;
  togglePlay: (invert?: boolean) => void;
  playerPaused: boolean;
  toggleDirection: () => void;
  direction: SpotifyPlayDirection;
  trackName?: string;
  artistName?: string;
  progressMs: number;
  durationMs: number;
  seekToPositionMs: (msTimeCode: number) => void;
};

type PlaybackControlButtonProps = {
  className: string;
  action: () => void;
  icon: IconDefinition;
  title: string;
  size?: "2x";
};

const PlaybackControlButton = (props: PlaybackControlButtonProps) => {
  const { className, action, icon, title, size } = props;
  return (
    <div
      className={className}
      title={title}
      onClick={action}
      onKeyPress={action}
      role="button"
      tabIndex={0}
    >
      <FontAwesomeIcon icon={icon as IconProp} size={size} />
    </div>
  );
};

type PlaybackControlsState = {
  autoHideControls: boolean;
};

export default class PlaybackControls extends React.Component<
  PlaybackControlsProps,
  PlaybackControlsState
> {
  // How many milliseconds to navigate to with keyboard left/right arrows
  keyboardStepMS: number = 5000;

  constructor(props: PlaybackControlsProps) {
    super(props);
    this.state = {
      autoHideControls: true,
    };
  }

  progressClickHandler = (event: React.MouseEvent<HTMLInputElement>): void => {
    const { durationMs, seekToPositionMs } = this.props;
    const progressBarBoundingRect = event.currentTarget.getBoundingClientRect();
    const progressBarWidth = progressBarBoundingRect.width;
    const musicPlayerXOffset = progressBarBoundingRect.x;
    const absoluteClickXPos = event.clientX;
    const relativeClickXPos = absoluteClickXPos - musicPlayerXOffset;
    const percentPos = relativeClickXPos / progressBarWidth;
    const positionMs = Math.round(durationMs * percentPos);
    seekToPositionMs(positionMs);
  };

  onKeyPressHandler = (event: React.KeyboardEvent<HTMLInputElement>): void => {
    const {
      durationMs,
      progressMs,
      seekToPositionMs,
      playNextTrack,
      playPreviousTrack,
    } = this.props;
    if (event.key === "ArrowLeft") {
      if (event.shiftKey) {
        playPreviousTrack();
        return;
      }
      const oneStepEarlier = progressMs - this.keyboardStepMS;
      seekToPositionMs(oneStepEarlier > 0 ? oneStepEarlier : 0);
    }
    if (event.key === "ArrowRight") {
      if (event.shiftKey) {
        playNextTrack();
        return;
      }
      const oneStepLater = progressMs + this.keyboardStepMS;
      if (oneStepLater <= durationMs - 500) {
        seekToPositionMs(oneStepLater);
      }
    }
  };

  render() {
    const {
      children,
      playerPaused,
      trackName,
      artistName,
      progressMs,
      durationMs,
      playPreviousTrack,
      togglePlay,
      playNextTrack,
      toggleDirection,
      direction,
    } = this.props;

    const { autoHideControls } = this.state;
    const progressPercentage = Number((progressMs * 100) / durationMs);
    return (
      <div id="music-player" aria-label="Playback control">
        <div className="album">
          {children || <div className="noAlbumArt">No album art</div>}
        </div>
        <div
          className={`info ${
            !autoHideControls || !children || playerPaused ? "showControls" : ""
          }`}
        >
          <div className="currently-playing">
            <h2 className="song-name">{trackName || "â€”"}</h2>
            <h3 className="artist-name">{artistName}</h3>
            <div
              className="progress"
              onClick={this.progressClickHandler}
              onKeyDown={this.onKeyPressHandler}
              aria-label="Audio Progress Control"
              role="progressbar"
              aria-valuemin={0}
              aria-valuemax={100}
              aria-valuenow={progressPercentage}
              tabIndex={0}
            >
              <div
                className="progress-bar"
                style={{
                  width: `${progressPercentage}%`,
                }}
              />
            </div>
          </div>
          <div className="controls">
            <PlaybackControlButton
              className="left btn btn-xs"
              title={`${
                autoHideControls ? "Always show" : "Autohide"
              } controls`}
              action={() => {
                this.setState((state) => ({
                  autoHideControls: !state.autoHideControls,
                }));
              }}
              icon={autoHideControls ? faEyeSlash : faEye}
            />
            <PlaybackControlButton
              className="previous btn btn-xs"
              title="Previous"
              action={playPreviousTrack}
              icon={faFastBackward}
            />
            <PlaybackControlButton
              className="play btn"
              action={togglePlay}
              title={`${playerPaused ? "Play" : "Pause"}`}
              icon={playerPaused ? faPlayCircle : faPauseCircle}
              size="2x"
            />
            <PlaybackControlButton
              className="next btn btn-xs"
              action={playNextTrack}
              title="Next"
              icon={faFastForward}
            />
            {direction !== "hidden" && (
              <PlaybackControlButton
                className="right btn btn-xs"
                action={toggleDirection}
                title={`Play ${direction === "up" ? "down" : "up"}`}
                icon={direction === "up" ? faSortAmountUp : faSortAmountDown}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
}
