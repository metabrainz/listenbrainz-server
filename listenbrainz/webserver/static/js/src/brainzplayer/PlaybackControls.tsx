import * as React from "react";
import {
  faCog,
  faEllipsisV,
  faEye,
  faEyeSlash,
  faFastBackward,
  faFastForward,
  faHeart,
  faHeartBroken,
  faPauseCircle,
  faPlayCircle,
} from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconDefinition } from "@fortawesome/fontawesome-common-types"; // eslint-disable-line import/no-unresolved
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { isNaN as _isNaN } from "lodash";

type PlaybackControlsProps = {
  playPreviousTrack: () => void;
  playNextTrack: (invert?: boolean) => void;
  togglePlay: (invert?: boolean) => void;
  playerPaused: boolean;
  trackName?: string;
  artistName?: string;
  progressMs: number;
  durationMs: number;
  seekToPositionMs: (msTimeCode: number) => void;
  currentFeedback: ListenFeedBack;
  submitFeedback: (score: ListenFeedBack) => Promise<void>;
};

type PlaybackControlButtonProps = {
  className: string;
  action: () => void;
  icon: IconDefinition;
  title: string;
  size?: "2x";
};

const PlaybackControlButton = (props: PlaybackControlButtonProps) => {
  const { className, action, icon, title, size } = props;
  return (
    <div
      className={className}
      title={title}
      onClick={action}
      onKeyPress={action}
      role="button"
      tabIndex={0}
    >
      <FontAwesomeIcon icon={icon as IconProp} size={size} />
    </div>
  );
};

type PlaybackControlsState = {};

export default class PlaybackControls extends React.Component<
  PlaybackControlsProps,
  PlaybackControlsState
> {
  // How many milliseconds to navigate to with keyboard left/right arrows
  keyboardStepMS: number = 5000;

  // constructor(props: PlaybackControlsProps) {
  //   super(props);
  //   this.state = {
  //   };
  // }

  progressClickHandler = (event: React.MouseEvent<HTMLInputElement>): void => {
    const { durationMs, seekToPositionMs } = this.props;
    const progressBarBoundingRect = event.currentTarget.getBoundingClientRect();
    const progressBarWidth = progressBarBoundingRect.width;
    const musicPlayerXOffset = progressBarBoundingRect.x;
    const absoluteClickXPos = event.clientX;
    const relativeClickXPos = absoluteClickXPos - musicPlayerXOffset;
    const percentPos = relativeClickXPos / progressBarWidth;
    const positionMs = Math.round(durationMs * percentPos);
    seekToPositionMs(positionMs);
  };

  onKeyPressHandler = (event: React.KeyboardEvent<HTMLInputElement>): void => {
    const {
      durationMs,
      progressMs,
      seekToPositionMs,
      playNextTrack,
      playPreviousTrack,
    } = this.props;
    if (event.key === "ArrowLeft") {
      if (event.shiftKey) {
        playPreviousTrack();
        return;
      }
      const oneStepEarlier = progressMs - this.keyboardStepMS;
      seekToPositionMs(oneStepEarlier > 0 ? oneStepEarlier : 0);
    }
    if (event.key === "ArrowRight") {
      if (event.shiftKey) {
        playNextTrack();
        return;
      }
      const oneStepLater = progressMs + this.keyboardStepMS;
      if (oneStepLater <= durationMs - 500) {
        seekToPositionMs(oneStepLater);
      }
    }
  };

  render() {
    const {
      children,
      playerPaused,
      trackName,
      artistName,
      progressMs,
      durationMs,
      playPreviousTrack,
      togglePlay,
      playNextTrack,
      currentFeedback,
      submitFeedback,
    } = this.props;

    const progressPercentage = Math.round(
      Number((progressMs * 100) / durationMs)
    );
    const isPlayingATrack = Boolean(trackName || artistName);
    const hideProgressBar =
      _isNaN(progressPercentage) || progressPercentage <= 0;
    return (
      <div id="brainz-player" aria-label="Playback control">
        <div
          className={`progress${hideProgressBar ? " hidden" : ""}`}
          onClick={this.progressClickHandler}
          onKeyDown={this.onKeyPressHandler}
          aria-label="Audio Progress Control"
          role="progressbar"
          aria-valuemin={0}
          aria-valuemax={100}
          aria-valuenow={progressPercentage}
          tabIndex={0}
        >
          <div
            className="progress-bar"
            style={{
              width: `${progressPercentage}%`,
            }}
          />
        </div>
        <div className="content">
          <div className="cover-art">
            {children}
            <div className="no-album-art" />
          </div>
          <div className={isPlayingATrack ? "currently-playing" : ""}>
            {trackName && (
              <div title={trackName} className="ellipsis-2-lines">
                {trackName}
              </div>
            )}
            {artistName && (
              <span className="small text-muted ellipsis" title={artistName}>
                {artistName}
              </span>
            )}
          </div>
        </div>
        <div className="controls">
          <PlaybackControlButton
            className="previous"
            title="Previous"
            action={playPreviousTrack}
            icon={faFastBackward}
          />
          <PlaybackControlButton
            className="play"
            action={togglePlay}
            title={`${playerPaused ? "Play" : "Pause"}`}
            icon={playerPaused ? faPlayCircle : faPauseCircle}
            size="2x"
          />
          <PlaybackControlButton
            className="next"
            action={playNextTrack}
            title="Next"
            icon={faFastForward}
          />
        </div>
        <div className="actions">
          <FontAwesomeIcon
            icon={faHeart}
            title="Love"
            onClick={() => submitFeedback(currentFeedback === 1 ? 0 : 1)}
            className={`${currentFeedback === 1 ? " loved" : ""}`}
          />
          <FontAwesomeIcon
            icon={faHeartBroken}
            title="Hate"
            onClick={() => submitFeedback(currentFeedback === -1 ? 0 : -1)}
            className={`${currentFeedback === -1 ? " hated" : ""}`}
          />
          {/* <FontAwesomeIcon icon={faEllipsisV} />
          <FontAwesomeIcon icon={faCog} /> */}
        </div>
      </div>
    );
  }
}
