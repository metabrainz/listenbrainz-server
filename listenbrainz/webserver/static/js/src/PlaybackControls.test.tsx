import * as React from "react";
import { mount } from "enzyme";

import PlaybackControls from "./PlaybackControls";

const props = {
  playPreviousTrack: () => {},
  playNextTrack: () => {},
  togglePlay: () => {},
  playerPaused: true,
  toggleDirection: () => {},
  direction: "up" as SpotifyPlayDirection,
  trackName: "Dangerous",
  artistName: "The xx",
  progressMs: 0,
  durationMs: 10000,
  seekToPositionMs: (msTimeCode: number) => {},
};
describe("PlaybackControls", () => {
  it("renders", () => {
    const wrapper = mount(<PlaybackControls {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
