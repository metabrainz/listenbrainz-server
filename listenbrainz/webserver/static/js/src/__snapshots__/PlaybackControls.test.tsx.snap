// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaybackControls renders 1`] = `
<PlaybackControls
  artistName="The xx"
  direction="up"
  durationMs={10000}
  playNextTrack={[Function]}
  playPreviousTrack={[Function]}
  playerPaused={true}
  progressMs={0}
  seekToPositionMs={[Function]}
  toggleDirection={[Function]}
  togglePlay={[Function]}
  trackName="Dangerous"
>
  <div
    aria-label="Playback control"
    id="music-player"
  >
    <div
      className="album"
    >
      <div
        className="noAlbumArt"
      >
        No album art
      </div>
    </div>
    <div
      className="info showControls"
    >
      <div
        className="currently-playing"
      >
        <h2
          className="song-name"
        >
          Dangerous
        </h2>
        <h3
          className="artist-name"
        >
          The xx
        </h3>
        <div
          aria-label="Audio Progress Control"
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="progress"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="progressbar"
          tabIndex={0}
        >
          <div
            className="progress-bar"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
      </div>
      <div
        className="controls"
      >
        <PlaybackControlButton
          action={[Function]}
          className="left btn btn-xs"
          icon={
            Object {
              "icon": Array [
                640,
                512,
                Array [],
                "f070",
                "M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z",
              ],
              "iconName": "eye-slash",
              "prefix": "fas",
            }
          }
          title="Always show controls"
        >
          <div
            className="left btn btn-xs"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Always show controls"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    640,
                    512,
                    Array [],
                    "f070",
                    "M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z",
                  ],
                  "iconName": "eye-slash",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-eye-slash fa-w-20 "
                data-icon="eye-slash"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </div>
        </PlaybackControlButton>
        <PlaybackControlButton
          action={[Function]}
          className="previous btn btn-xs"
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f049",
                "M0 436V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v151.9L235.5 71.4C256.1 54.3 288 68.6 288 96v131.9L459.5 71.4C480.1 54.3 512 68.6 512 96v320c0 27.4-31.9 41.7-52.5 24.6L288 285.3V416c0 27.4-31.9 41.7-52.5 24.6L64 285.3V436c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12z",
              ],
              "iconName": "fast-backward",
              "prefix": "fas",
            }
          }
          title="Previous"
        >
          <div
            className="previous btn btn-xs"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Previous"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f049",
                    "M0 436V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v151.9L235.5 71.4C256.1 54.3 288 68.6 288 96v131.9L459.5 71.4C480.1 54.3 512 68.6 512 96v320c0 27.4-31.9 41.7-52.5 24.6L288 285.3V416c0 27.4-31.9 41.7-52.5 24.6L64 285.3V436c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12z",
                  ],
                  "iconName": "fast-backward",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-fast-backward fa-w-16 "
                data-icon="fast-backward"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 436V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v151.9L235.5 71.4C256.1 54.3 288 68.6 288 96v131.9L459.5 71.4C480.1 54.3 512 68.6 512 96v320c0 27.4-31.9 41.7-52.5 24.6L288 285.3V416c0 27.4-31.9 41.7-52.5 24.6L64 285.3V436c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </div>
        </PlaybackControlButton>
        <PlaybackControlButton
          action={[Function]}
          className="play btn"
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f144",
                "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z",
              ],
              "iconName": "play-circle",
              "prefix": "fas",
            }
          }
          size="2x"
          title="Play"
        >
          <div
            className="play btn"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Play"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f144",
                    "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z",
                  ],
                  "iconName": "play-circle",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size="2x"
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-play-circle fa-w-16 fa-2x "
                data-icon="play-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </div>
        </PlaybackControlButton>
        <PlaybackControlButton
          action={[Function]}
          className="next btn btn-xs"
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f050",
                "M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z",
              ],
              "iconName": "fast-forward",
              "prefix": "fas",
            }
          }
          title="Next"
        >
          <div
            className="next btn btn-xs"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Next"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f050",
                    "M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z",
                  ],
                  "iconName": "fast-forward",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-fast-forward fa-w-16 "
                data-icon="fast-forward"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </div>
        </PlaybackControlButton>
        <PlaybackControlButton
          action={[Function]}
          className="right btn btn-xs"
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f161",
                "M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.38-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.77 160 16 160zm416 0H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
              ],
              "iconName": "sort-amount-up",
              "prefix": "fas",
            }
          }
          title="Play down"
        >
          <div
            className="right btn btn-xs"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title="Play down"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f161",
                    "M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.38-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.77 160 16 160zm416 0H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
                  ],
                  "iconName": "sort-amount-up",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-sort-amount-up fa-w-16 "
                data-icon="sort-amount-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.38-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.77 160 16 160zm416 0H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </div>
        </PlaybackControlButton>
      </div>
    </div>
  </div>
</PlaybackControls>
`;
